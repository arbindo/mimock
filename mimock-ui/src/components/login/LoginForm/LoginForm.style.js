import tw from 'tailwind-styled-components';
import styled from 'styled-components';
import { CustomButton } from 'styles/Button';

export const LoginFormContainer = tw.div`
    w-11/12
    h-full
    flex
    flex-col
    mx-auto
    my-auto
    justify-center
    align-middle
`;

export const Title = tw.div`
    font-sans
    text-xl
    font-bold
    text-gray-700
`;

export const Underline = tw.div`
    bg-teal-500
    w-36
    h-1.5
`;

export const Label = tw.div`
    font-sans
    text-gray-500
    mt-3
`;

export const InputWrapper = tw.form`
    w-11/12
    my-5
    mx-auto
    flex
    flex-col
    justify-center
`;

export const UserNameField = tw.input`
    w-full
    my-2
    p-2
    outline-none
    rounded-md
    border-2
    border-gray-300
`;

export const PasswordField = styled(UserNameField)``;

export const SubmitButton = tw(CustomButton)``;

export const ErrorLabel = tw.div`
    font-sans
    rounded-md
    my-2
    mx-auto
    text-center
    p-3
    shadow-sm
    bg-red-100
    text-red-700
`;
