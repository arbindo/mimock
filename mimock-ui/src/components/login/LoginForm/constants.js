export const constants = {
	labels: {
		userName: 'USER NAME',
		password: 'PASSWORD',
	},
	placeholders: {
		userName: 'Enter username',
		password: 'Enter password',
	},
	errors: {
		userNameEmpty: 'Username cannot be empty',
		passwordEmpty: 'Password cannot be empty',
		loginFailed: 'User login failed. Please try again.',
	},
};
