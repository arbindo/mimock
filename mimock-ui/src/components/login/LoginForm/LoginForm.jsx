import React, { useState } from 'react';
import { Cookies } from 'react-cookie';
import { useNavigate } from 'react-router-dom';
import { getToken } from 'api/MimockApi';
import {
	LoginFormContainer,
	Title,
	Underline,
	InputWrapper,
	Label,
	UserNameField,
	PasswordField,
	SubmitButton,
	ErrorLabel,
} from './LoginForm.style';
import { constants } from './constants';

export default function LoginForm() {
	const cookies = new Cookies();
	const navigate = useNavigate();

	const [errorMessage, setErrorMessage] = useState('');
	const [userName, setUserName] = useState('');
	const [password, setPassword] = useState('');

	const isCredentialsValid = () => {
		if (userName === '') {
			setErrorMessage(constants.errors.userNameEmpty);
			return false;
		}

		if (password === '') {
			setErrorMessage(constants.errors.passwordEmpty);
			return false;
		}

		return true;
	};

	const resetError = () => {
		setErrorMessage('');
	};

	const authenticateUser = async () => {
		if (!isCredentialsValid()) {
			return;
		}

		await getToken(userName, password)
			.then((res) => {
				const { token } = res.data;
				cookies.set('__authToken', token);
				navigate('/mocks');
			})
			.catch(() => {
				setErrorMessage(constants.errors.loginFailed);
			});
	};

	return (
		<LoginFormContainer data-testid='login-form'>
			<Title>Login to mimock</Title>
			<Underline />

			<InputWrapper
				onSubmit={(e) => {
					e.preventDefault();
					resetError();
					authenticateUser();
				}}
			>
				<Label data-testid='login-username-label'>
					{constants.labels.userName}
				</Label>
				<UserNameField
					data-testid='login-username-input'
					placeholder={constants.placeholders.userName}
					required
					autoComplete='off'
					value={userName}
					onChange={(e) => {
						resetError();
						setUserName(e.currentTarget.value);
					}}
				/>

				<Label data-testid='login-password-label'>
					{constants.labels.password}
				</Label>
				<PasswordField
					data-testid='login-password-input'
					type='password'
					placeholder={constants.placeholders.password}
					required
					autoComplete='off'
					value={password}
					onChange={(e) => {
						resetError();
						setPassword(e.currentTarget.value);
					}}
				/>

				<SubmitButton
					dataTestid='login-submit'
					background='bg-teal-500'
					label='Login'
					color='text-white'
					width='w-full'
					additionalStyles='mt-5'
					type='submit'
				/>
			</InputWrapper>
			<If condition={errorMessage}>
				<ErrorLabel data-testid='login-error-label'>{errorMessage}</ErrorLabel>
			</If>
		</LoginFormContainer>
	);
}
