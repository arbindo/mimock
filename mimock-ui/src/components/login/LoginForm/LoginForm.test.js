import React from 'react';
import LoginForm from './LoginForm';
import { render, fireEvent, act } from '@testing-library/react';

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => {
	return {
		useNavigate: () => mockNavigate,
	};
});

describe('LoginForm', () => {
	afterEach(() => {
		jest.clearAllMocks();
	});

	it('should render login form', async () => {
		const tree = await render(<LoginForm />);

		const { container, getByTestId, queryByTestId } = tree;

		expect(getByTestId('login-username-label')).toBeInTheDocument();
		expect(getByTestId('login-username-label').textContent).toStrictEqual(
			'USER NAME'
		);

		expect(getByTestId('login-password-label')).toBeInTheDocument();
		expect(getByTestId('login-password-label').textContent).toStrictEqual(
			'PASSWORD'
		);

		expect(getByTestId('login-username-input')).toBeInTheDocument();
		expect(getByTestId('login-password-input')).toBeInTheDocument();

		expect(getByTestId('login-submit')).toBeInTheDocument();

		expect(queryByTestId('login-error-label')).not.toBeInTheDocument();

		expect(container).toMatchSnapshot();
	});

	it.skip('should enter credentials to login', async () => {
		const tree = await render(<LoginForm />);

		const { container, getByTestId, queryByTestId } = tree;

		await act(async () => {
			const userName = getByTestId('login-username-input');
			await fireEvent.change(userName, {
				target: {
					value: 'mimock_user',
				},
			});
			expect(userName.value).toBe('mimock_user');

			const password = getByTestId('login-password-input');
			await fireEvent.change(password, {
				target: {
					value: 'test@123',
				},
			});
			expect(password.value).toBe('test@123');
		});

		const loginBtn = getByTestId('login-submit');
		await fireEvent.click(loginBtn);

		expect(mockNavigate()).toHaveBeenCalledTimes(1);
		expect(mockNavigate).toHaveBeenCalledWith('/mocks');

		expect(queryByTestId('login-error-label')).not.toBeInTheDocument();

		expect(container).toMatchSnapshot();
	});
});
