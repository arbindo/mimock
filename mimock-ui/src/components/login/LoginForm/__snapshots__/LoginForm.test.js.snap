// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should enter credentials to login 1`] = `
<div>
  <div
    class="w-11/12 h-full flex flex-col mx-auto my-auto justify-center align-middle"
    data-testid="login-form"
  >
    <div
      class="font-sans text-xl font-bold text-gray-700"
    >
      Login to mimock
    </div>
    <div
      class="bg-teal-500 w-36 h-1.5"
    />
    <form
      class="w-11/12 my-5 mx-auto flex flex-col justify-center"
    >
      <div
        class="font-sans text-gray-500 mt-3"
        data-testid="login-username-label"
      >
        USER NAME
      </div>
      <input
        autocomplete="off"
        class="w-full my-2 p-2 outline-none rounded-md border-2 border-gray-300"
        data-testid="login-username-input"
        placeholder="Enter username"
        required=""
        value="mimock_user"
      />
      <div
        class="font-sans text-gray-500 mt-3"
        data-testid="login-password-label"
      >
        PASSWORD
      </div>
      <input
        autocomplete="off"
        class="w-full my-2 p-2 outline-none rounded-md border-2 border-gray-300 sc-bdvvtL"
        data-testid="login-password-input"
        placeholder="Enter password"
        required=""
        type="password"
        value="test@123"
      />
      <button
        class="px-9 py-2 mx-auto shadow-md rounded-md text-center text-base font-semibold cursor-pointer hover:brightness-110 mt-5 text-white w-full bg-teal-500"
        data-testid="login-submit"
      >
        Login
      </button>
    </form>
  </div>
</div>
`;

exports[`LoginForm should render login form 1`] = `
<div>
  <div
    class="w-11/12 h-full flex flex-col mx-auto my-auto justify-center align-middle"
    data-testid="login-form"
  >
    <div
      class="font-sans text-xl font-bold text-gray-700"
    >
      Login to mimock
    </div>
    <div
      class="bg-teal-500 w-36 h-1.5"
    />
    <form
      class="w-11/12 my-5 mx-auto flex flex-col justify-center"
    >
      <div
        class="font-sans text-gray-500 mt-3"
        data-testid="login-username-label"
      >
        USER NAME
      </div>
      <input
        autocomplete="off"
        class="w-full my-2 p-2 outline-none rounded-md border-2 border-gray-300"
        data-testid="login-username-input"
        placeholder="Enter username"
        required=""
        value=""
      />
      <div
        class="font-sans text-gray-500 mt-3"
        data-testid="login-password-label"
      >
        PASSWORD
      </div>
      <input
        autocomplete="off"
        class="w-full my-2 p-2 outline-none rounded-md border-2 border-gray-300 sc-bdvvtL"
        data-testid="login-password-input"
        placeholder="Enter password"
        required=""
        type="password"
        value=""
      />
      <button
        class="px-9 py-2 mx-auto shadow-md rounded-md text-center text-base font-semibold cursor-pointer hover:brightness-110 mt-5 text-white w-full bg-teal-500"
        data-testid="login-submit"
      >
        Login
      </button>
    </form>
  </div>
</div>
`;

exports[`LoginForm should show error message credentials are empty 1`] = `
<div>
  <div
    class="w-11/12 h-full flex flex-col mx-auto my-auto justify-center align-middle"
    data-testid="login-form"
  >
    <div
      class="font-sans text-xl font-bold text-gray-700"
    >
      Login to mimock
    </div>
    <div
      class="bg-teal-500 w-36 h-1.5"
    />
    <form
      class="w-11/12 my-5 mx-auto flex flex-col justify-center"
    >
      <div
        class="font-sans text-gray-500 mt-3"
        data-testid="login-username-label"
      >
        USER NAME
      </div>
      <input
        autocomplete="off"
        class="w-full my-2 p-2 outline-none rounded-md border-2 border-gray-300"
        data-testid="login-username-input"
        placeholder="Enter username"
        required=""
        value=""
      />
      <div
        class="font-sans text-gray-500 mt-3"
        data-testid="login-password-label"
      >
        PASSWORD
      </div>
      <input
        autocomplete="off"
        class="w-full my-2 p-2 outline-none rounded-md border-2 border-gray-300 sc-bdvvtL"
        data-testid="login-password-input"
        placeholder="Enter password"
        required=""
        type="password"
        value=""
      />
      <button
        class="px-9 py-2 mx-auto shadow-md rounded-md text-center text-base font-semibold cursor-pointer hover:brightness-110 mt-5 text-white w-full bg-teal-500"
        data-testid="login-submit"
      >
        Login
      </button>
    </form>
    <div
      class="font-sans rounded-md my-2 mx-auto text-center p-3 shadow-sm bg-red-100 text-red-700"
      data-testid="login-error-label"
    >
      Username cannot be empty
    </div>
  </div>
</div>
`;

exports[`LoginForm should show error message when getToken api call fails 1`] = `
<div>
  <div
    class="w-11/12 h-full flex flex-col mx-auto my-auto justify-center align-middle"
    data-testid="login-form"
  >
    <div
      class="font-sans text-xl font-bold text-gray-700"
    >
      Login to mimock
    </div>
    <div
      class="bg-teal-500 w-36 h-1.5"
    />
    <form
      class="w-11/12 my-5 mx-auto flex flex-col justify-center"
    >
      <div
        class="font-sans text-gray-500 mt-3"
        data-testid="login-username-label"
      >
        USER NAME
      </div>
      <input
        autocomplete="off"
        class="w-full my-2 p-2 outline-none rounded-md border-2 border-gray-300"
        data-testid="login-username-input"
        placeholder="Enter username"
        required=""
        value="mimock_user"
      />
      <div
        class="font-sans text-gray-500 mt-3"
        data-testid="login-password-label"
      >
        PASSWORD
      </div>
      <input
        autocomplete="off"
        class="w-full my-2 p-2 outline-none rounded-md border-2 border-gray-300 sc-bdvvtL"
        data-testid="login-password-input"
        placeholder="Enter password"
        required=""
        type="password"
        value="test@123"
      />
      <button
        class="px-9 py-2 mx-auto shadow-md rounded-md text-center text-base font-semibold cursor-pointer hover:brightness-110 mt-5 text-white w-full bg-teal-500"
        data-testid="login-submit"
      >
        Login
      </button>
    </form>
    <div
      class="font-sans rounded-md my-2 mx-auto text-center p-3 shadow-sm bg-red-100 text-red-700"
      data-testid="login-error-label"
    >
      User login failed. Please try again.
    </div>
  </div>
</div>
`;

exports[`LoginForm should show error message when password is empty 1`] = `
<div>
  <div
    class="w-11/12 h-full flex flex-col mx-auto my-auto justify-center align-middle"
    data-testid="login-form"
  >
    <div
      class="font-sans text-xl font-bold text-gray-700"
    >
      Login to mimock
    </div>
    <div
      class="bg-teal-500 w-36 h-1.5"
    />
    <form
      class="w-11/12 my-5 mx-auto flex flex-col justify-center"
    >
      <div
        class="font-sans text-gray-500 mt-3"
        data-testid="login-username-label"
      >
        USER NAME
      </div>
      <input
        autocomplete="off"
        class="w-full my-2 p-2 outline-none rounded-md border-2 border-gray-300"
        data-testid="login-username-input"
        placeholder="Enter username"
        required=""
        value="mimock_user"
      />
      <div
        class="font-sans text-gray-500 mt-3"
        data-testid="login-password-label"
      >
        PASSWORD
      </div>
      <input
        autocomplete="off"
        class="w-full my-2 p-2 outline-none rounded-md border-2 border-gray-300 sc-bdvvtL"
        data-testid="login-password-input"
        placeholder="Enter password"
        required=""
        type="password"
        value=""
      />
      <button
        class="px-9 py-2 mx-auto shadow-md rounded-md text-center text-base font-semibold cursor-pointer hover:brightness-110 mt-5 text-white w-full bg-teal-500"
        data-testid="login-submit"
      >
        Login
      </button>
    </form>
    <div
      class="font-sans rounded-md my-2 mx-auto text-center p-3 shadow-sm bg-red-100 text-red-700"
      data-testid="login-error-label"
    >
      Password cannot be empty
    </div>
  </div>
</div>
`;

exports[`LoginForm should show error message when user name is empty 1`] = `
<div>
  <div
    class="w-11/12 h-full flex flex-col mx-auto my-auto justify-center align-middle"
    data-testid="login-form"
  >
    <div
      class="font-sans text-xl font-bold text-gray-700"
    >
      Login to mimock
    </div>
    <div
      class="bg-teal-500 w-36 h-1.5"
    />
    <form
      class="w-11/12 my-5 mx-auto flex flex-col justify-center"
    >
      <div
        class="font-sans text-gray-500 mt-3"
        data-testid="login-username-label"
      >
        USER NAME
      </div>
      <input
        autocomplete="off"
        class="w-full my-2 p-2 outline-none rounded-md border-2 border-gray-300"
        data-testid="login-username-input"
        placeholder="Enter username"
        required=""
        value=""
      />
      <div
        class="font-sans text-gray-500 mt-3"
        data-testid="login-password-label"
      >
        PASSWORD
      </div>
      <input
        autocomplete="off"
        class="w-full my-2 p-2 outline-none rounded-md border-2 border-gray-300 sc-bdvvtL"
        data-testid="login-password-input"
        placeholder="Enter password"
        required=""
        type="password"
        value="test@123"
      />
      <button
        class="px-9 py-2 mx-auto shadow-md rounded-md text-center text-base font-semibold cursor-pointer hover:brightness-110 mt-5 text-white w-full bg-teal-500"
        data-testid="login-submit"
      >
        Login
      </button>
    </form>
    <div
      class="font-sans rounded-md my-2 mx-auto text-center p-3 shadow-sm bg-red-100 text-red-700"
      data-testid="login-error-label"
    >
      Username cannot be empty
    </div>
  </div>
</div>
`;
